### PR TITLE
Accessibility Enhancements

### PR DESCRIPTION
**Pull request by jonathan343**
## Overview
This PR provides two enhancements related to accessibility for the botocore/boto3 documentation websites.
1. **Keyboard Navigation** - Keyboard only users currently cant open/close the left or right sidebars since the related input label icons cannot be tabbed. This PR makes these labels (and other relevant ones) accessible through tabbing and adds new close icons to provide keyboard-only users a mechanism to open and close these sidebars.
2. **Implicit Headings** - A previous PR attempted to address some of these issues by keeping a list of `STRONG_TO_H3_HEADINGS` texts value we wanted to address. Since then, we've learned that we're responsible for converting more unpredictable values. This PR replaces the old method with a new one that uses the common pattern of `<p><strong>Implicit Headings</strong><p>` to detect and replace with `<h3>Implicit Heading</h3>`.

## Solutions
**Keyboard Navigation**
1. Add labels that can be used for closing sidebars.
2. Make sure all open/close labels are only visible when necessary. 
  > Note: There are two different labels/icons for opening the right sidebar (for the table of contents). One thats displayed 
on medium sized screens and one thats displayed on smaller screens. This comes from Furo.
3. Enable tabbing for all labels with `tabindex="0"`
4. Activate input when a label is in focus and a user clicks "Enter".
5. The correct label should be put in focus after opening or closing a sidebar.
  * Example: When we open the right side bar, the close icon should be focused on. When we close the sidebar, the open 
   icon should be focused on. Same should happen for the left side bar. This feature is demonstrated in the section below.

**Implicit Headings**
1. To detect the pattern `<p><strong>Implicit Headings</strong><p>`
  * To do this we look for paragraph nodes who have only 1 child with the type `docutil.nodes.strong`. This child node must also only have 1 child node of the type `docutil.nodes.Text`. The text value of this node should not exist in our `IGNORE_IMPLICIT_HEADINGS` list.
2. When a node with the pattern above is detected, store the text value of the `docutil.nodes.Text` node and write this as an `h3` tag instead.
3. `raise nodes.SkipNode` to skip visiting the current nodes children and its call of the depart method.

## Demo
demo

### CODE REVIEW

File: docs/source/_templates/page.html
Comment (jonathan343) [inline docs/source/_templates/page.html:1]: This screenshot provides a diff when compared to the original [page.html](https://github.com/pradyunsg/furo/blob/main/src/furo/theme/furo/page.html) from furo.
screenshot

File: botocore/docs/translator.py [hunk 1]
```diff
@@ -39,24 +33,32 @@ def visit_admonition(self, node, name=""):
...
             self.body.append(title)
 
-    def visit_strong(self, node):
-        """Visit a strong HTML element.
+    def is_implicit_heading(self, node):
+        """Determines if a node is an implicit heading.
 
-        Opens the h3 tag for a specific set of words/phrases and opens the
-        strong tag for all others.
+        An implicit heading is represented by a paragraph node whos only
```
Comment (dlm6693) [inline botocore/docs/translator.py:39]:
```suggestion
        An implicit heading is represented by a paragraph node whose only
```

File: botocore/docs/translator.py [hunk 2]
```diff
@@ -39,24 +33,32 @@ def visit_admonition(self, node, name=""):
...
-            self.body.append(self.starttag(node, "strong", ""))
+        if (
+            len(node) == 1
+            and isinstance(node[0], nodes.strong)
+            and len(node[0]) == 1
+            and isinstance(node[0][0], nodes.Text)
+            and node[0][0].astext() not in self.IGNORE_IMPLICIT_HEADINGS
+        ):
+            return True
+        return False
```
Comment (dlm6693) [inline botocore/docs/translator.py:50]:
```suggestion
        return (
            len(node) == 1
            and isinstance(node[0], nodes.strong)
            and len(node[0]) == 1
            and isinstance(node[0][0], nodes.Text)
            and node[0][0].astext() not in self.IGNORE_IMPLICIT_HEADINGS
        )
```

File: docs/source/_templates/page.html [hunk 3]
```diff
@@ -11,110 +11,243 @@
...
-        </div>
-        {% if prev.link == pathto(master_doc) %}
-        <div class="title">{{ _("Home") }}</div>
-        {% else %}
-        <div class="title">{{ prev.title }}</div>
-        {% endif %}
-        </div>
-    </a>
-    {%- endif %}
+{% block body -%}
```
[the remark this answers is not in the document]
Comment (jonathan343) [reply docs/source/_templates/page.html:14]: So in furo's original [page.html](https://github.com/pradyunsg/furo/blob/main/src/furo/theme/furo/page.html) the `footer` block is nested in the `body` block. Previously, we only needed to modify the `footer` so utilizing jinja2's [template inheritance](https://jinja.palletsprojects.com/en/3.0.x/templates/#template-inheritance) we could just do:
```
{%- extends "!page.html" %}

{% block footer %}
    # Custom HTML that overrides furo's footer block
{%- endblock %}
```
This uses all the original `body` block content but replaces the `footer` block with what we defined. Ideally, as many things would be wrapped in its own block to make things more easily customizable without having to rewrite an entire template. In this case we need to modify elements in the body (outside of the footer), so we need to now do something like:
```
{%- extends "!page.html" %}
{% block body %}
    # Custom HTML before footer
    {% block footer %}
        # Custom HTML that overrides furo's footer block
    {%- endblock %}
    # Custom HTML after footer
 {%- endblock %}
```

File: docs/source/_templates/sidebar/close-icon.html [hunk 4]
```diff
@@ -0,0 +1,6 @@
+<div class="sidebar-div show-div-sm justify-content-right">
+    <label class="nav-close-icon" id="nav-menu-close" for="__navigation" tabindex="0">
+        <div class="visually-hidden">Toggle site navigation sidebar</div>
+        <i class="icon"><svg><use href="#svg-close"></use></svg></i>
```
[the remark this answers is not in the document]
Comment (jonathan343) [reply docs/source/_templates/sidebar/close-icon.html:4]: I was mainly trying to stay as consistent as possible with what Furo does. They have an [icons.html](https://github.com/pradyunsg/furo/blob/main/src/furo/theme/furo/partials/icons.html) page where they define different symbols and later include this in [page.html](https://github.com/pradyunsg/furo/blob/main/src/furo/theme/furo/page.html#L5). There is no easy way to add symbols to this list so I added the new symbol directly [here](https://github.com/boto/botocore/pull/2949/files#diff-744d1d98c2fcc6c08f7cede4744a333ec07932af4be05a17df36dca5c394aab5R19-R28). 

This seems like valid behavior according to [svgwg.org](https://svgwg.org/svg2-draft/struct.html#UseElement)

File: docs/source/_templates/sidebar/close-icon.html
[the remark this answers is not in the document]
Comment (jonathan343) [reply docs/source/_templates/sidebar/close-icon.html:1]: This file gets references in [conf.py](https://github.com/boto/botocore/pull/2949/files#diff-008dcb3426febd767787b1521f1fe33086313b927ea37eaab86df5fa88a51698R172). More information on customizing the left sidebar can be read on [furo's doc site](http://127.0.0.1:49177/customisation/sidebar.html#making-changes).

File: docs/source/_static/css/custom.css [hunk 5]
```diff
@@ -31,7 +31,59 @@ h1, code, li {
...
+}
+@media (max-width: 82em) {
+     /* Displays a div on a medium screen. */
+    .show-div-md {
+        display: flex;
+    }
+    /* Hides the right sidebar on a medium screen by default. */
+    .toc-drawer {
+        display: none;
+    }
```
[the remark this answers is not in the document]
Comment (jonathan343) [reply docs/source/_static/css/custom.css:81]: No, but the way furo currently does this is by animating the left/right side bar outside the visible screen area. The issue with this is that keyboard users can continue to tab/focus on elements they cant see. Applying the `display: none;` style prevents this behavior. The downside of this is that we loose the cool animation. Im looking into a way to achieve both.

File: botocore/docs/translator.py [hunk 6]
```diff
@@ -39,24 +33,32 @@ def visit_admonition(self, node, name=""):
...
-            self.body.append(self.starttag(node, "h3", ""))
-        else:
-            self.body.append(self.starttag(node, "strong", ""))
+        if (
+            len(node) == 1
+            and isinstance(node[0], nodes.strong)
+            and len(node[0]) == 1
+            and isinstance(node[0][0], nodes.Text)
+            and node[0][0].astext() not in self.IGNORE_IMPLICIT_HEADINGS
+        ):
```
[the remark this answers is not in the document]
Comment (jonathan343) [reply botocore/docs/translator.py:48]: This logic happens after. #2855 applied changes in the transition from `upstream HTML --> RST`. This translator extends the one from sphinx that is used in the transition from `RST --> Boto Docs HTML`.

File: docs/source/_templates/page.html
Comment (jonathan343) [reply docs/source/_templates/page.html:1]: Updated the image with a new diff. I think i solved the issue you're talking about? It was extra whitespace and newlines.

**Review by nateprewitt — APPROVED** [review]
Cool, this all looks great. Thanks for the thorough deep dive on this @jonathan343 and it's exciting we may be able to port some of this back upstream.

My only question is how does that transition look if it is released in a future version of furo. Do we get that for free and can revert this or will we need to keep specific parts in the Translator/CSS/Page Template?

**Comment by jonathan343** [comment]
> My only question is how does that transition look if it is released in a future version of furo. Do we get that for free and can revert this or will we need to keep specific parts in the Translator/CSS/Page Template?

@nateprewitt If https://github.com/pradyunsg/furo/pull/660 is ever merged and released in furo, we'll be able to revert the parts related to keyboard navigation (everything except changes to `translator.py`) since they'll be part of furo.

**Pull request merged**